lib/datadog/ci/ext/environment.rb

```diff
@@ -312,6 +317,107 @@ def extract_bitrise(env)
+          Datadog.logger.debug("Unable to read git commit users: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_repository_url
+          exec_git_command('git ls-remote --get-url')
+        rescue => e
+          Datadog.logger.debug("Unable to read git repository url: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_commit_message
+          exec_git_command('git show -s --format=%s')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit message: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_branch
+          exec_git_command('git rev-parse --abbrev-ref HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git branch: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_commit_sha
+          exec_git_command('git rev-parse HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit SHA: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_tag
+          exec_git_command('git tag --points-at HEAD')
+        rescue => e
+          Datadog.logger.debug("Unable to read git commit SHA: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def git_base_directory
+          git_dir = exec_git_command('git rev-parse --git-dir')
+          return unless git_dir
+
+          # The result will point to the '.git' directory.
+          # We actually want the base source directory.
+          File.expand_path('..', git_dir)
+        rescue => e
+          Datadog.logger.debug("Unable to read git dir path: #{e.message} at #{e.backtrace.first}")
+          nil
+        end
+
+        def exec_git_command(cmd)
+          result = `#{cmd} 2>/dev/null`
+          return nil if result.empty?
+
+          result.strip # There's always a "\n" at the end of the command output
+        end
+
+        def extract_local_git
+          author, committer = git_commit_users
+
+          {
+            TAG_WORKSPACE_PATH => git_base_directory,
+            Datadog::Ext::Git::TAG_REPOSITORY_URL => git_repository_url,
+            Datadog::Ext::Git::TAG_COMMIT_SHA => git_commit_sha,
+            Datadog::Ext::Git::TAG_BRANCH => git_branch,
+            Datadog::Ext::Git::TAG_TAG => git_tag,
+            Datadog::Ext::Git::TAG_COMMIT_MESSAGE => git_commit_message,
+            Datadog::Ext::Git::TAG_COMMIT_AUTHOR_NAME => (author[:name] if author),
+            Datadog::Ext::Git::TAG_COMMIT_AUTHOR_EMAIL => (author[:email] if author),
+            Datadog::Ext::Git::TAG_COMMIT_AUTHOR_DATE => (author[:date] if author),
+            Datadog::Ext::Git::TAG_COMMIT_COMMITTER_NAME => (committer[:name] if committer),
+            Datadog::Ext::Git::TAG_COMMIT_COMMITTER_EMAIL => (committer[:email] if committer),
+            Datadog::Ext::Git::TAG_COMMIT_COMMITTER_DATE => (committer[:date] if committer)
```

Comment:
Minor: This may be simplified a bit if rather than an optional array of optional hashes, we got just a flat hash from `git_commit_users` with `{author_name: ..., author_email: ...}`. It's a small change, but it's a nicer and easier to work with structure that reduces [connascence](https://en.wikipedia.org/wiki/Connascence).

Reply:
I wrote it like this at first (`{author_name: ..., author_email: ...}`), and it looked more confusing to me, but I can try reverting it again and seeing how it looks now with the final version.

Reply:
I refactored this bit, and I think it looks easier to read now :)